# Raw Data Preparation

## [Description of Each Field Can Also Be Found Here](https://data.boston.gov/dataset/property-assessment/resource/96b6cf8b-04f2-4b87-b78f-ba9a0ddd1573)

<details>
<summary><strong>🏠 Property Identification</strong></summary>

- **PID**: Unique 10‐digit parcel number. First 2 digits are the ward, digits 3 to 7 are the parcel, and digits 8 to 10 are the sub‐parcel
- **CM_ID**: 10‐digit parcel number of Condo Main, which houses all related condo units
- **GIS_ID**: Primary GIS ID
- **BLDG_SEQ**: Building sequence of parcel
- **NUM_BLDGS**: Total number of buildings of parcel

</details>

<details>
<summary><strong>📍 Location & Address</strong></summary>

- **ST_NUM**: Street number of parcel
- **ST_NAME**: Street name of parcel
- **UNIT_NUM**: Specific unit number within a housing complex
- **CITY**: City or Town of parcel
- **ZIP_CODE**: Zip code of parcel

</details>

<details>
<summary><strong>🏘️ Property Classification</strong></summary>

- **LUC**: Land Use Code
  - *[Detailed Description of Each Code - See Documentation Above](#luc-documentation)*
  - We are going to focus our attention on the Residential Property Only. The full list of the residential codes is:
    - **101**: SINGLE FAM DWELLING
    - **102**: RESIDENTIAL CONDO
    - **103**: MOBILE HOME
    - **104**: TWO-FAM DWELLING
    - **105**: THREE-FAM DWELLING
    - **106**: ADD'L RES IMPROVEMENT
    - **107**: OTHER RESIDENTIAL
    - **108**: CONDO PARKING
    - **109**: MULTIPLE BUILDINGS
    - **110**: CONDO STORAGE

- **LU**: Type of Property (land use)
  - **A**: Residential 7 or more units
  - **AH**: Agricultural/Horticultural
  - **C**: Commercial
  - **CC**: Commercial condominium
  - **CD**: Residential condominium unit
  - **CL**: Commercial land
  - **CM**: Condominium main (physical structure housing all related condo units with no assessed value)
  - **CP**: Condo parking
  - **E**: Tax‐exempt
  - **EA**: Tax‐exempt (121A)
  - **I**: Industrial
  - **R1**: Residential 1‐family
  - **R2**: Residential 2‐family
  - **R3**: Residential 3‐family
  - **R4**: Residential 4 or more family
  - **RC**: Mixed use (residential and commercial)
  - **RL**: Residential land

- **LU_DESC**: Land Use Description

</details>

<details>
<summary><strong>👤 Ownership & Mailing</strong></summary>

- **OWN_OCC**: Residential Exemption: "Y" character code indicating if owner receives residential exemption as an owner‐occupied property
<div style="background-color: #ffebee; border: 2px solid #f44336; border-radius: 5px; padding: 10px; margin: 5px 0;">
<span style="color: #d32f2f; font-weight: bold;">⚠️ PII VIOLATION WARNING:</span> 
<strong>OWNER</strong> field contains personally identifiable information and should NEVER be published publicly or used in modeling. I deleted this field prior to loading this data to GitHub.
</div>

- **MAIL_ADDRESSEE**: Care of recipient
- **MAIL_ADDRESS**: Street address where tax bill is mailed
- **MAIL_CITY**: City/neighborhood where tax bill is mailed
- **MAIL_STATE**: State where tax bill is mailed
- **MAIL_ZIP_CODE**: Zip code where tax bill is mailed

</details>

<details>
<summary><strong>📏 Building Size & Layout</strong></summary>

- **RES_FLOOR**: Number of residential building stories
- **CD_FLOOR**: Condominium unit floor number
- **RES_UNITS**: Number of residential units in a condominium building
- **COM_UNITS**: Number of commercial units in a condominium building
- **RC_UNITS**: Number of Residential/Commercial Units in a condominium building
- **LAND_SF**: Parcel's land area in square feet (legal area)
- **GROSS_AREA**: Gross floor area
- **LIVING_AREA**: Living area square footage of the property

</details>

<details>
<summary><strong>💰 Property Values & Taxes</strong></summary>

- **LAND_VALUE**: Total assessed land value
- **BLDG_VALUE**: Total assessed building value
- **GROSS_TAX**: Tax bill amount based on total assessed value multiplied by the tax rate per thousand

</details>

<details>
<summary><strong>🏗️ Building Characteristics & Age</strong></summary>

- **BLDG_TYPE**: Building Type and Style
- **YR_BUILT**: Year property was built
- **YR_REMODEL**: Year property was last remodeled
- **STRUCTURE_CL**: Structural classification of building
  - **A**: Struct Steel
  - **B**: Reinforced Concrete
  - **C**: Brick/Concrete
  - **D**: Wood/Frame
  - **E**: Metal
  - **R**: Residential

</details>

<details>
<summary><strong>🏠 Exterior Features</strong></summary>

- **ROOF_STRUCTURE**: Roof Structure Types
  - **F**: Flat
  - **G**: Gable
  - **H**: Hip
  - **L**: Gambrel
  - **M**: Mansard
  - **O**: Other
  - **S**: Shed

- **ROOF_COVER**: Roof Cover Material
  - **A**: Asphalt Shingle
  - **C**: Composition
  - **O**: Other
  - **R**: Rubber Roof
  - **S**: Slate
  - **T**: Tile
  - **W**: Wood Shingle

- **EXT_FINISHED**: Exterior Siding Material

</details>

<details>
<summary><strong>🔧 Property Condition</strong></summary>

- **INT_WALL**: Interior Wall Condition‐Residential
- **INT_COND**: Interior Condition of parcel‐Residential
- **EXT_COND**: Exterior Condition of parcel
- **OVERALL_COND**: Overall condition of parcel
- **BDRM_COND**: Bedroom Condition‐Residential

</details>

<details>
<summary><strong>🛏️ Rooms & Living Spaces</strong></summary>

- **BED_RMS**: Total number of bedrooms‐Residential
- **FULL_BTH**: Total number of full baths‐Residential
- **HALF_BTH**: Total number of half baths‐Residential
- **KITCHEN**: Total number of kitchens‐Residential
- **TT_RMS**: Total number of rooms‐Residential

</details>

<details>
<summary><strong>🚿 Bathroom Details</strong></summary>

- **BTHRM_STYLE1**: Residential bath style ‐bathroom #1
- **BTHRM_STYLE2**: Residential bath style ‐bathroom #2
- **BTHRM_STYLE3**: Residential bath style ‐ bathroom #3

</details>

<details>
<summary><strong>🍳 Kitchen Details</strong></summary>

- **KITCHEN_TYPE**: Kitchen Type ‐ Residential
- **KITCHEN_STYLE1**: Residential kitchen style –kitchen #1
- **KITCHEN_STYLE2**: Residential kitchen style – kitchen #2
- **KITCHEN_STYLE3**: Residential kitchen style –kitchen #3

</details>

<details>
<summary><strong>🌡️ Heating & Cooling Systems</strong></summary>

- **HEAT_TYPE**: Heating type
- **HEAT_SYSTEM**: Heating System (individually controlled, common or self‐contained)
- **AC_TYPE**: Air Conditioning Type‐Residential

</details>

<details>
<summary><strong>⭐ Amenities & Special Features</strong></summary>

- **FIRE_PLACE**: Total number of fireplaces
- **NUM_PARKING**: Number of parking spaces
- **PROP_VIEW**: Property View
- **ORIENTATION**: Indicates the location a condominium unit in the building is facing
- **CORNER_UNIT**: Indicates if the unit is locating in the corner of the building (Y/N)

</details>

## Getting the Data

In [46]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")
df = pd.read_csv('fy2025-property-assessment-data_12_30_2024.csv', dtype={'ZIP_CODE': str})

In [47]:
rows, cols = df.shape
print(f"The dataset contains {rows:,} records with {cols} columns")

The dataset contains 183,445 records with 65 columns


### Column Analysis

In [48]:
# Clean the columns
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')

In [49]:
df.columns

Index(['pid', 'cm_id', 'gis_id', 'st_num', 'st_num2', 'st_name', 'unit_num',
       'city', 'zip_code', 'bldg_seq', 'num_bldgs', 'luc', 'lu', 'lu_desc',
       'bldg_type', 'own_occ', 'mail_addressee', 'mail_street_address',
       'mail_city', 'mail_state', 'mail_zip_code', 'res_floor', 'cd_floor',
       'res_units', 'com_units', 'rc_units', 'land_sf', 'gross_area',
       'living_area', 'land_value', 'bldg_value', 'sfyi_value', 'total_value',
       'gross_tax', 'yr_built', 'yr_remodel', 'structure_class',
       'roof_structure', 'roof_cover', 'int_wall', 'ext_fnished', 'int_cond',
       'ext_cond', 'overall_cond', 'bed_rms', 'full_bth', 'hlf_bth',
       'kitchens', 'tt_rms', 'bdrm_cond', 'bthrm_style1', 'bthrm_style2',
       'bthrm_style3', 'kitchen_type', 'kitchen_style1', 'kitchen_style2',
       'kitchen_style3', 'heat_type', 'heat_system', 'ac_type', 'fireplaces',
       'orientation', 'num_parking', 'prop_view', 'corner_unit'],
      dtype='object')

### Renaming Columns

In [50]:
# Column renaming dictionary
column_rename_map = {
    'pid': 'parcel_id',
    'cm_id': 'condominium_id', 
    'gis_id': 'geographic_id',
    'st_num': 'street_num',
    'st_num2': 'street_num2',
    'st_name': 'street_name',
    'bldg_seq': 'bldg_sequence',
    'lu': 'land_use_type',
    'luc': 'land_use_code',
    'lu_desc': 'land_use_description',
    'own_occ': 'owner_occupied',
    'res_floor': 'residential_floor',
    'cd_floor': 'condo_floor', 
    'res_units': 'residential_units',
    'com_units': 'commercial_units',
    'rc_units': 'res_or_commercial_units',
    'hlf_bth': 'half_bth',
    'tt_rms': 'total_num_rooms',
    'prop_view': 'property_view'
}

# Apply the renaming
df = df.rename(columns=column_rename_map)

In [51]:
df.head(4)

,parcel_id,condominium_id,geographic_id,street_num,street_num2,street_name,unit_num,city,zip_code,bldg_sequence,...,kitchen_style2,kitchen_style3,heat_type,heat_system,ac_type,fireplaces,orientation,num_parking,property_view,corner_unit
0,100001000,NaN,100001000,104.0,NaN,PUTNAM ST,NaN,EAST BOSTON,02128,1,...,S - Semi-Modern,S - Semi-Modern,W - Ht Water/Steam,NaN,N - None,0.0,NaN,3.0,A - Average,NaN
1,100002000,NaN,100002000,197.0,NaN,Lexington ST,NaN,EAST BOSTON,02128,1,...,M - Modern,M - Modern,F - Forced Hot Air,NaN,C - Central AC,0.0,NaN,0.0,A - Average,NaN
2,100003000,NaN,100003000,199.0,NaN,Lexington ST,NaN,EAST BOSTON,02128,1,...,S - Semi-Modern,S - Semi-Modern,S - Space Heat,NaN,N - None,0.0,NaN,0.0,A - Average,NaN
3,100004000,NaN,100004000,201.0,NaN,Lexington ST,NaN,EAST BOSTON,02128,1,...,S - Semi-Modern,S - Semi-Modern,W - Ht Water/Steam,NaN,N - None,0.0,NaN,0.0,A - Average,NaN


In [52]:
df.columns

Index(['parcel_id', 'condominium_id', 'geographic_id', 'street_num',
       'street_num2', 'street_name', 'unit_num', 'city', 'zip_code',
       'bldg_sequence', 'num_bldgs', 'land_use_code', 'land_use_type',
       'land_use_description', 'bldg_type', 'owner_occupied', 'mail_addressee',
       'mail_street_address', 'mail_city', 'mail_state', 'mail_zip_code',
       'residential_floor', 'condo_floor', 'residential_units',
       'commercial_units', 'res_or_commercial_units', 'land_sf', 'gross_area',
       'living_area', 'land_value', 'bldg_value', 'sfyi_value', 'total_value',
       'gross_tax', 'yr_built', 'yr_remodel', 'structure_class',
       'roof_structure', 'roof_cover', 'int_wall', 'ext_fnished', 'int_cond',
       'ext_cond', 'overall_cond', 'bed_rms', 'full_bth', 'half_bth',
       'kitchens', 'total_num_rooms', 'bdrm_cond', 'bthrm_style1',
       'bthrm_style2', 'bthrm_style3', 'kitchen_type', 'kitchen_style1',
       'kitchen_style2', 'kitchen_style3', 'heat_type', 'heat_

### Filtering for Residential Units Only

In [53]:
# focus on residential properties only. 
residential_list = ['CD', 'R1', 'R2', 'R3', 'R4', 'RC']

- A: Residential 7 or more units
- CD: Residential condominium unit
- R1: Residential 1‐family
- R2: Residential 2‐family
- R3: Residential 3‐family
- R4: Residential 4 or more family
- RC: Mixed use (residential and commercial)

In [54]:
df_residential = df[df['land_use_type'].isin(residential_list)]

In [55]:
len(df_residential)

139463

In [56]:
df_residential.head(3)

,parcel_id,condominium_id,geographic_id,street_num,street_num2,street_name,unit_num,city,zip_code,bldg_sequence,...,kitchen_style2,kitchen_style3,heat_type,heat_system,ac_type,fireplaces,orientation,num_parking,property_view,corner_unit
0,100001000,NaN,100001000,104.0,NaN,PUTNAM ST,NaN,EAST BOSTON,02128,1,...,S - Semi-Modern,S - Semi-Modern,W - Ht Water/Steam,NaN,N - None,0.0,NaN,3.0,A - Average,NaN
1,100002000,NaN,100002000,197.0,NaN,Lexington ST,NaN,EAST BOSTON,02128,1,...,M - Modern,M - Modern,F - Forced Hot Air,NaN,C - Central AC,0.0,NaN,0.0,A - Average,NaN
2,100003000,NaN,100003000,199.0,NaN,Lexington ST,NaN,EAST BOSTON,02128,1,...,S - Semi-Modern,S - Semi-Modern,S - Space Heat,NaN,N - None,0.0,NaN,0.0,A - Average,NaN


### Checking Missing Values

In [57]:
df_residential.isna().sum().to_dict()

{'parcel_id': 0,
 'condominium_id': 65944,
 'geographic_id': 0,
 'street_num': 280,
 'street_num2': 125341,
 'street_name': 0,
 'unit_num': 65964,
 'city': 0,
 'zip_code': 0,
 'bldg_sequence': 0,
 'num_bldgs': 0,
 'land_use_code': 0,
 'land_use_type': 0,
 'land_use_description': 0,
 'bldg_type': 46,
 'owner_occupied': 0,
 'mail_addressee': 117937,
 'mail_street_address': 9,
 'mail_city': 7,
 'mail_state': 266,
 'mail_zip_code': 27,
 'residential_floor': 3048,
 'condo_floor': 66086,
 'residential_units': 139463,
 'commercial_units': 139463,
 'res_or_commercial_units': 139463,
 'land_sf': 73,
 'gross_area': 419,
 'living_area': 425,
 'land_value': 0,
 'bldg_value': 0,
 'sfyi_value': 0,
 'total_value': 0,
 'gross_tax': 0,
 'yr_built': 417,
 'yr_remodel': 58774,
 'structure_class': 131882,
 'roof_structure': 5373,
 'roof_cover': 5366,
 'int_wall': 5367,
 'ext_fnished': 421,
 'int_cond': 5368,
 'ext_cond': 5311,
 'overall_cond': 7031,
 'bed_rms': 5367,
 'full_bth': 95,
 'half_bth': 91,
 'ki

### Examing One Record 

In [58]:
df_residential.iloc[14804].to_dict()

{'parcel_id': 300475419,
 'condominium_id': 300475000.0,
 'geographic_id': 300475000,
 'street_num': 9.0,
 'street_num2': nan,
 'street_name': 'Hawthorne PL',
 'unit_num': '7-K',
 'city': 'BOSTON',
 'zip_code': '02114',
 'bldg_sequence': 1,
 'num_bldgs': 1,
 'land_use_code': 102,
 'land_use_type': 'CD',
 'land_use_description': 'RESIDENTIAL CONDO',
 'bldg_type': 'HR - High Rise',
 'owner_occupied': 'N',
 'mail_addressee': 'C/O PAUL WU',
 'mail_street_address': '567 LOWELL ST',
 'mail_city': 'LEXINGTON',
 'mail_state': 'MA',
 'mail_zip_code': '02420',
 'residential_floor': 1.0,
 'condo_floor': 7.0,
 'residential_units': nan,
 'commercial_units': nan,
 'res_or_commercial_units': nan,
 'land_sf': '545',
 'gross_area': 637.0,
 'living_area': 637.0,
 'land_value': '0',
 'bldg_value': '434,200',
 'sfyi_value': '0',
 'total_value': '434,200',
 'gross_tax': '$5,028.04 ',
 'yr_built': 1967.0,
 'yr_remodel': 2007.0,
 'structure_class': nan,
 'roof_structure': 'F - Flat',
 'roof_cover': 'C - Comp

### Getting Rid of Irrelevant Columns 

- Columns to be removed are the ones that pertain to the mailing address of the owner (fields that start with `_mail`)
- Bathroom Style (`bhthrm_style`) and kitchen style (`kitchen_style`) columns are to be removed since overall condition (`overall_cond`) field is available
- Bedroom Condition (`bdrm_cond`), Interior Condition of parcel‐Residential (`int_cond`), Exterior condition of parcel (`ext_cond`), and Interior Wall Condition‐Residential (`int_wall`) columns will also be removed since overall condition (`overall_cond`) field is available and will allow for a simpler modeling
- Total assessed land value (`land_value`) column is to be removed because we are dealing with residential properties and not land
- Number of Residential/Commercial Units in a condominium building (`res_or_commercial_units`)  can either increase value (convenience, rental income potential) or decrease it (noise, traffic, commercial activity), making the signal unclear. Also, the analysis targets individual residential property values - what matters is the specific unit's characteristics, not the building's overall commercial mix
- Number of commercial units in a condominium building (`commercial_units`) since the analysis targets residential property buyers and investors, the number of commercial units in a building doesn't impact residential unit valuation decisions
- Number of residential units in a condominium building (`residential_units`) since the assumption is that one unit is examed at a time
- Columns that contain information about condominium id (`condominium_id`), parcel id (`parcel_id`) and Geographic information system id (`geographic_id`) since the information about zip code (`zip_code`) is available
- Anything about street number(s) (`street_num` and `street_num2`), street name (`street_name`), unit number (`unit_num`), and city (`city`) since the information about zip code (`zip_code`) is available
- `sfyi_value` - this field has not been described at all. If the field is not described, it would not be wise to use it
- Structure Class (`structure_class`) not only has 131882 fields missing but also structure class is highly correlated with year built - older buildings are typically wood frame, newer ones use steel/concrete. Also, The construction material itself doesn't directly determine value - what matters is the resulting condition, durability, and maintenance costs (captured by condition fields)
- Building Sequence (`bldg_sequence`) only applies to multi-building parcels, making it irrelevant for most residential properties and the sequence number itself doesn't influence what buyers are willing to pay
- Total number of buildings of parcel (`num_bldgs`) is excluded as well since most residential parcels have only one building, creating minimal variation
- Total assessed building value (`bldg_value`) is not relevant since the the target value of `total_value` has been selected
- Parcel's land area in square feet (legal area) (`land_sf`) and living area square footage of the property (`living_area`) are not relevant since the field Gross floor area (`gross_area`) is available and should be used for modeling. It is also the field that is actually used for the tax purposes.

- Exterior Siding Material (`ext_fnished`) is not relevant because siding type affects maintenance costs but has minimal impact on property value in Boston's mature market. Also, in a historic city like Boston, architectural character matters more than specific siding materials.
- Corner unit (`corner_unit`) is a limited-scope field being applicable only to condominiums, excluding single-family homes and other property types in the scope of the analysis
- Property view (`property_view`) is not relevant because the property view is subjective to the property assessor's view. People who are going to purchase the property might have a completely different view. 
- Roof Cover Material (`roof_cover`) and Roof Structure Types (`roof_structure`) are also not very applicable in the case of purchase. The only feature that would impact the purchase of the property is how recently the roof was renovated. This field is not available in the dataset. 
- The location a condominium unit in the building is facing (`orientation`) should also be excluded because this field only applies to condominium units, excluding single-family homes, two-family homes, and other residential property types that make up a significant portion of our dataset
- Heating System (individually controlled, common or self‐contained) (`heat_system`) has 66081 and there is already information available about heat with far less missing data - Heating type (`heat_type`)
- Kitchen Type ‐ Residential (`kitchen_type`) since there is a numeric representation of kitchens, which is under the field `kitchen`

In [59]:
unwanted_columns = (
    [col for col in df_residential.columns if col.startswith('mail_')] + 
    [col for col in df_residential.columns if col.startswith('bthrm_style')] +
    [col for col in df_residential.columns if col.startswith('kitchen_style')] +
    ['bdrm_cond', 'int_cond', 'ext_cond', 'int_wall',  'land_value', 'res_or_commercial_units', 'commercial_units', 'residential_units']+
    ['condominium_id', 'parcel_id', 'geographic_id', 'street_num', 'street_num2', 'street_name', 'unit_num', 'city'] +
    ['sfyi_value', 'structure_class', 'bldg_sequence', 'num_bldgs', 'bldg_value', 'land_sf', 'living_area'] +
    ['ext_fnished', 'corner_unit', 'property_view', 'roof_cover' , 'roof_structure', 'orientation', 'heat_system', 'kitchen_type']
     
)

df_dropped_columns = df_residential.drop(columns=unwanted_columns)

In [60]:
df_dropped_columns.head(3)

,zip_code,land_use_code,land_use_type,land_use_description,bldg_type,owner_occupied,residential_floor,condo_floor,gross_area,total_value,...,overall_cond,bed_rms,full_bth,half_bth,kitchens,total_num_rooms,heat_type,ac_type,fireplaces,num_parking
0,02128,105,R3,THREE-FAM DWELLING,RE - Row End,N,3.0,NaN,3353.0,"799,000",...,A - Average,6.0,3.0,0.0,3.0,12.0,W - Ht Water/Steam,N - None,0.0,3.0
1,02128,105,R3,THREE-FAM DWELLING,RM - Row Middle,N,3.0,NaN,3047.0,"843,000",...,A - Average,3.0,3.0,0.0,3.0,9.0,F - Forced Hot Air,C - Central AC,0.0,0.0
2,02128,105,R3,THREE-FAM DWELLING,RM - Row Middle,Y,3.0,NaN,3392.0,"815,900",...,A - Average,5.0,3.0,0.0,3.0,13.0,S - Space Heat,N - None,0.0,0.0


### Filtering Mixed use (residential and commercial) Properties (RC) to Only Those with Room Data

In [61]:
df_dropped_columns_filtered = df_dropped_columns[~((df_dropped_columns['land_use_type'] == 'RC') & (df_dropped_columns['total_num_rooms'].isna()))]

In [62]:
len(df_dropped_columns_filtered)

136591

In [63]:
df_dropped_columns_filtered.loc[14804].to_dict()

{'zip_code': '02129',
 'land_use_code': 102,
 'land_use_type': 'CD',
 'land_use_description': 'RESIDENTIAL CONDO',
 'bldg_type': 'FS - Free Standing',
 'owner_occupied': 'Y',
 'residential_floor': 1.0,
 'condo_floor': 3.0,
 'gross_area': 789.0,
 'total_value': '568,200',
 'gross_tax': '$6,579.75 ',
 'yr_built': 1878.0,
 'yr_remodel': 2004.0,
 'overall_cond': nan,
 'bed_rms': 1.0,
 'full_bth': 1.0,
 'half_bth': 0.0,
 'kitchens': 1.0,
 'total_num_rooms': 3.0,
 'heat_type': 'W - Ht Water/Steam',
 'ac_type': 'C - Central AC',
 'fireplaces': 0.0,
 'num_parking': 0.0}

In [64]:
df_dropped_columns_filtered.shape

(136591, 23)

### Saving the simplified version of the dataset as a csv for further cleaning, EDA, feature engineering and modeling

In [65]:
df_dropped_columns_filtered.to_csv('uncleaned_property_data.csv', index = False)